File: strands-py/src/strands/models/routing/fallback_strategy.py
"""The default routing strategy: ordered failover that prefers the candidates failing least."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .strategy import RoutingContext, _attempts_since_success

if TYPE_CHECKING:
    from .router import RoutingCandidate


class FallbackStrategy:
    """Picks the healthiest candidate not yet tried since the last success.

    Candidates already tried since the last success are excluded, then the fewest recorded failures
    wins, ties going to the earlier declaration. So an invocation with no failures behind it is plain
    declaration order, and a model that keeps failing sinks below healthier ones rather than being
    re-tried in its declared slot. A success clears the failures recorded before it, which both
    re-arms an earlier candidate for a later round and stops that success from being demoted.
    Returns ``None`` once every candidate has been tried since the last success.
    """

    async def select(self, context: RoutingContext, **kwargs: Any) -> RoutingCandidate | None:
        """Return the least-failed candidate not yet tried since the last success, else ``None``."""
        failures: dict[int, int] = {}
        for attempt in context.attempts:
            if attempt.exception is None:
                failures.pop(id(attempt.candidate), None)
            else:
                failures[id(attempt.candidate)] = failures.get(id(attempt.candidate), 0) + 1

        tried_now = {id(attempt.candidate) for attempt in _attempts_since_success(context.attempts)}
        available = [candidate for candidate in context.candidates if id(candidate) not in tried_now]
        if not available:
            return None
        return min(available, key=lambda candidate: failures.get(id(candidate), 0))

File: strands-ts/src/models/routing/fallback-strategy.ts
import type { RoutingCandidate } from './router.js'
import type { RoutingAttempt, RoutingContext, RoutingStrategy } from './strategy.js'

/** Selects the healthiest candidate not yet tried since the last success. */
export class FallbackStrategy implements RoutingStrategy {
  /**
   * Select the least-failed available candidate, breaking ties by declaration order.
   *
   * @param context - Current routing context
   * @returns The selected candidate, or `undefined` when the round is exhausted
   */
  async select(context: RoutingContext): Promise<RoutingCandidate | undefined> {
    const failures = new Map<RoutingCandidate, number>()
    for (const attempt of context.attempts) {
      if (attempt.exception === undefined) {
        failures.delete(attempt.candidate)
      } else {
        failures.set(attempt.candidate, (failures.get(attempt.candidate) ?? 0) + 1)
      }
    }

    const triedNow = new Set(attemptsSinceSuccess(context.attempts).map((attempt) => attempt.candidate))
    const available = context.candidates.filter((candidate) => !triedNow.has(candidate))
    if (available.length === 0) return undefined

    return available.reduce((best, candidate) =>
      (failures.get(candidate) ?? 0) < (failures.get(best) ?? 0) ? candidate : best
    )
  }
}

function attemptsSinceSuccess(attempts: readonly RoutingAttempt[]): readonly RoutingAttempt[] {
  for (let index = attempts.length - 1; index >= 0; index -= 1) {
    if (attempts[index]!.exception === undefined) return attempts.slice(index + 1)
  }
  return attempts
}
